### PR TITLE
fix(material-experimental/theming): Make M3 work with typography-hierarchy

### PR DESCRIPTION
**Pull request by mmalerba**

### CODE REVIEW

File: src/material/core/typography/_typography.scss [hunk 1]
```diff
@@ -8,11 +11,133 @@
...
+@mixin typography-hierarchy($theme, $selector: '.mat-typography', $back-compat: false) {
+  @if inspection.get-theme-version($theme) == 1 {
+    @include _m3-typography-hierarchy($theme, $selector, $back-compat);
+  }
+  @else {
+    @include _m2-typography-hierarchy($theme, $selector);
+  }
+}
+
+@function _get-selector($selectors, $prefix) {
```
Comment (zarend) [inline src/material/core/typography/_typography.scss:23]: Question: do you think there's a way to unit test to verify we don't have "naked" tag selectors, nested prefix selector, etc.?

File: src/material/core/typography/_typography.scss [hunk 2]
```diff
@@ -8,11 +11,133 @@
...
+    @if $add-m2-selectors {
+      $options: map.set($options, selectors,
+          list.join(map.get($options, selectors), map.get($options, m2-selectors) or ()));
+    }
+    $options: map.remove($options, m2-selectors);
+
+    // Apply styles for the level.
+    @include _m3-typography-level($theme, $selector-prefix, $level, $options...);
+
+    // Also style <p> inside body-large.
```
Comment (zarend) [inline src/material/core/typography/_typography.scss:125]: I'm wondering why paragraph element is treated as a special case. Do you think we will eventually need to handle more types of elements like pre, blockquote, figure, etc.?

**Review by zarend — APPROVED** [review]
Looks reasonable to me, left some questions.

File: src/material/core/typography/_typography.scss (hunk 2 above)
Comment (mmalerba) [reply src/material/core/typography/_typography.scss:125]: This is just maintaining the existing behavior of typography-hierarchy for M2 themes. We could add others in the future if we decide, but there are no immediate plans to.

File: src/material/core/typography/_typography.scss (hunk 1 above)
Comment (mmalerba) [reply src/material/core/typography/_typography.scss:23]: I can probably write a test that looks at the CSS output and verifies this

File: src/material/core/theming/tests/theming-typography-hierarchy.spec.ts [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+        css,
+        '.mat-display-large, .mat-typography .mat-display-large, .mat-typography h1',
+      );
+    });
+
+    it('should emit default body styles', () => {
+      const css = transpile('@include mat.typography-hierarchy($theme)');
+      verifyFullSelector(css, '.mat-body-large, .mat-typography .mat-body-large, .mat-typography');
+    });
+
```
Comment (crisbeto) [inline src/material/core/theming/tests/theming-typography-hierarchy.spec.ts:67]: Should we have a test where the selector is included under another class?

**Review by crisbeto — APPROVED** [review]

File: src/material/core/typography/_typography.scss (hunk 1 above)
Comment (mmalerba) [reply src/material/core/typography/_typography.scss:23]: Done

File: src/material/core/theming/tests/theming-typography-hierarchy.spec.ts (hunk 3 above)
Comment (mmalerba) [reply src/material/core/theming/tests/theming-typography-hierarchy.spec.ts:67]: Done

**Pull request merged**